src/parser/jobs/dnc/modules/DirtyDancing.tsx

```diff
@@ -0,0 +1,298 @@
+import {t} from '@lingui/macro'
+import {Plural, Trans} from '@lingui/react'
+import _ from 'lodash'
+import React, {Fragment} from 'react'
+import {Icon, Message} from 'semantic-ui-react'
+
+import {ActionLink, StatusLink} from 'components/ui/DbLink'
+import {RotationTable} from 'components/ui/RotationTable'
+import ACTIONS from 'data/ACTIONS'
+import STATUSES from 'data/STATUSES'
+import {CastEvent, DamageEvent} from 'fflogs'
+import Module, {dependency} from 'parser/core/Module'
+import CheckList, {Requirement, Rule} from 'parser/core/modules/Checklist'
+import Combatants from 'parser/core/modules/Combatants'
+import Invulnerability from 'parser/core/modules/Invulnerability'
+import Suggestions, {SEVERITY, TieredSuggestion} from 'parser/core/modules/Suggestions'
+import Timeline from 'parser/core/modules/Timeline'
+
+const ISSUE_SEVERITY_TIERS = {
+	1: SEVERITY.MINOR,
+	3: SEVERITY.MEDIUM,
+	5: SEVERITY.MAJOR,
+}
+
+const STEPS = [
+	ACTIONS.STANDARD_STEP.id,
+	ACTIONS.TECHNICAL_STEP.id,
+]
+
+const DANCE_MOVES = [
+	ACTIONS.ENTRECHAT.id,
+	ACTIONS.EMBOITE.id,
+	ACTIONS.JETE.id,
+	ACTIONS.PIROUETTE.id,
+]
+
+const STANDARD_FINISHES = [
+	ACTIONS.STANDARD_FINISH.id,
+	ACTIONS.SINGLE_STANDARD_FINISH.id,
+	ACTIONS.DOUBLE_STANDARD_FINISH.id,
+]
+
+const TECHNICAL_FINISHES = [
+	ACTIONS.TECHNICAL_FINISH.id,
+	ACTIONS.SINGLE_TECHNICAL_FINISH.id,
+	ACTIONS.DOUBLE_TECHNICAL_FINISH.id,
+	ACTIONS.TRIPLE_TECHNICAL_FINISH.id,
+	ACTIONS.QUADRUPLE_TECHNICAL_FINISH.id,
+]
+
+const FINISHES = [
+	...STANDARD_FINISHES,
+	...TECHNICAL_FINISHES,
+]
+
+const EXPECTED_DANCE_MOVE_COUNT = {
+	[ACTIONS.DOUBLE_STANDARD_FINISH.id]: 2,
+	[ACTIONS.QUADRUPLE_TECHNICAL_FINISH.id]: 4,
+}
+
+class Dance {
+	start: number
+	end?: number
+	rotation: CastEvent[] = []
+	dancing: boolean = false
+	resolved: boolean = false
+
+	dirty: boolean = false
+	missed: boolean = false
+	footloose: boolean = false
+
+	public get error(): boolean {
+		return this.dirty || this.missed || this.footloose
+	}
+
+	public get expectedFinishId(): number {
+		const actualFinish = _.last(this.rotation)
+		let expectedFinish = -1
+		if (actualFinish) {
```

Comment:
nit; can invert some of these and shortcut with `return` to flatten it out a bunch

Reply:
I can't think of a satisfactory way to flatten this one that a) keeps the .includes checks from firing if actualFinish is undefined, and b) doesn't have a bunch of extra conditionals to make sure we only return a real value if the actual finisher is contained in TECHNICAL/STANDARD_FINISHES. That might be 4 hours of sleep talking though so I'll look at it again tomorrow unless this gets merged first.

Reply:
Can fix it later, but something like this usually works:

```ts
if (!actualFinish) {
	return -1
}

if (tech) {
	return quad_finish
}

if (standard) {
	return double_finish
}
```